### PR TITLE
[FSDP][state_dict][3/N] Change how state_dict utils access attributes in _FSDPState

### PR DESCRIPTION
**Pull request by fegin**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #88638
* #88637
* #88636
* __->__ #88635
* #88481
* #87900

**What This PR Does**
_state_dict_utils currently accesses the FSDP states through module. To enable composable FSDP state_dict, these accesses need to go through _FSDPState. module is still required for most APIs as state_dict has to access per-module information.

### CODE REVIEW

File: torch/distributed/fsdp/_common_utils.py [hunk 1]
```diff
@@ -61,6 +61,24 @@ def _all_handles(state: _FSDPState) -> List:
     )
 
 
+@no_type_check
+def _module_handles(module: nn.Module, state: _FSDPState) -> List:
+    """
+    Given a module and returns the flat handles that map to this module. If the
+    module is FullyShardedDataParallel, the module._handles will be returned.
+    """
```
Comment (awgu) [inline torch/distributed/fsdp/_common_utils.py:69]: I am wondering what you think about using the `_is_composable()` util from `_common_utils.py` instead of `if module == state`.

I can see that `if module == state` encodes the intention explicitly, but at the same time, I can see that enforcing the call to `_is_composable(state)` each time the logic between the two code paths diverges can make it easier to find such divergences (i.e. just search for `_is_composable`).

Basically, the implementation of `_is_composable(state)` performs a weaker check of simply seeing if `state` is an instance of `nn.Module`, but it still might be a good idea to use that check instead. Let me know what you think!

File: torch/distributed/fsdp/_common_utils.py [hunk 2]
```diff
@@ -61,6 +61,24 @@ def _all_handles(state: _FSDPState) -> List:
     )
 
 
+@no_type_check
+def _module_handles(module: nn.Module, state: _FSDPState) -> List:
```
Comment (awgu) [inline torch/distributed/fsdp/_common_utils.py:65]: nit: I am also wondering how you feel about argument ordering. I recall that every function/method I refactored had `state: _FSDPState` as the first argument. I do not think this makes any difference functionally, but maybe we keep that convention?

File: torch/distributed/fsdp/_state_dict_utils.py [hunk 3]
```diff
@@ -617,13 +672,16 @@ def _pre_load_state_dict_hook(
...
-        fsdp_module, state_dict, prefix
+    _pre_load_state_dict_hook_fn[fsdp_state._state_dict_type](
+        fsdp_module, fsdp_state, state_dict, prefix
     )
 
 
+@no_type_check
 @torch.no_grad()
 def _post_load_state_dict_hook(module: nn.Module, *args: Any) -> None:
+    # TODO: get the composable state from module
```
Comment (awgu) [inline torch/distributed/fsdp/_state_dict_utils.py:683]: Just FYI, I think the syntax is like:
```
fsdp_state = fully_shard.state(module)
```

For example:
https://github.com/pytorch/pytorch/blob/fca6ed02b91e0d685cc9b8b504bac5d356d31876/torch/distributed/_composable/fully_shard.py#L44

**Review by awgu — APPROVED** [review]
LGTM!

File: torch/distributed/fsdp/_common_utils.py (hunk 1 above)
Comment (fegin) [reply torch/distributed/fsdp/_common_utils.py:69]: Sure, let's use `is_composable(state)`.

File: torch/distributed/fsdp/_state_dict_utils.py (hunk 3 above)
Comment (fegin) [reply torch/distributed/fsdp/_state_dict_utils.py:683]: hmm, this can cause a circular import. Can we pull this to a module that does not depend on FSDP?

Comment (awgu) [reply torch/distributed/fsdp/_state_dict_utils.py:683]: Discussed over chat -- let us discuss this further with the composability devs
